### PR TITLE
Versioning

### PR DESCRIPTION
**Pull request by ahgittin**
docs re cleared-up OSGi/semver/maven versioning, as per ML

(this is WIP because the corresponding code is not yet written; however review here could be useful)

### CODE REVIEW

File: guide/blueprints/catalog/bundle.md [hunk 1]
```diff
@@ -5,7 +5,25 @@ layout: website-normal
...
+when you have multiple blueprints that you want to keep in sync. Brooklyn will persist any 
+uploaded bundles so that they are available after a restart or on HA failover.
+
+The bundle must be a ZIP file including a `catalog.bom` in the root.
+(The `br` CLI will create a ZIP from a local folder, for convenience.)
+The `catalog.bom` must declare a `bundle` identifier and a `version`, 
+following Brooklyn's [versioning](versioning.html) rules.
+Brooklyn will keep track of that bundle, allow SNAPSHOT-version bundles to be replaced,
+and ensure dependent bundles (specified as `brooklyn.libraries` or, for people familiar
+with OSGi, the `Required-bundle` manifest header) are available for searching.
```
Comment (drigodwin) [inline guide/blueprints/catalog/bundle.md]: I would probably remove `for searching` as it's not clear what searching is in this context.

File: guide/blueprints/catalog/versioning.md [hunk 2]
```diff
@@ -3,18 +3,118 @@ title: Versioning
...
 
-### Versioning
+Brooklyn supports multiple versions of a type to be installed and used at the same time.
+Versions are a first-class concept and are often prominently displayed in the UI.
 
-Version numbers follow the OSGi convention. This can have a major, minor, micro and qualifier part.
-For example, `1.0`. `1.0.1` or `1.0.1-20150101`.
+In order to do this, Brooklyn requires that the `id:version` string be unique across the catalog:
+it is normally an error to add a type if a type with the same `id:version` is present.
+Exceptions are if the definition is identical, or if the `version` is noted as a `SNAPSHOT`.
```
Comment (drigodwin) [inline guide/blueprints/catalog/versioning.md]: Exceptions are -> The exceptions to this are

File: guide/blueprints/catalog/versioning.md [hunk 3]
```diff
@@ -3,18 +3,118 @@ title: Versioning
...
+you will see `1.10.0.rc3-20170619`.
+This mapping is guaranteed to be one-to-one so no conflicts will occur if the
+recommended syntax is used.
+(If not using the recommended syntax, the mapping proceeds by treating the first dot-separated fragment 
+as the qualifer and converts unsupported characters in a qualifier to an underscore;
+thus `1.x` becomes `1.0.0.x`, `v1` becomes `0.0.0.v1`, and `"1.0.0-v1.1"` becomes `"1.0.0.v1_1"` 
+hence the bundle replacement noted above.)
+
+If you are creating an OSGi `MANIFEST.MF` for a bundle that also contains a `catalog.bom`, 
+you will need to use the mapped rsult (OSGi version syntax) in the manifest,
```
Comment (drigodwin) [inline guide/blueprints/catalog/versioning.md]: rsult -> result

**Comment by ahgittin** [comment]
Thanks @drigodwin, all addressed.  Should not be merged until after https://github.com/apache/brooklyn-server/pull/743 .

Also added explanation about the `0` minor/patch versions as per @aledsage 's ML comment.

File: guide/blueprints/catalog/versioning.md [hunk 4]
```diff
@@ -3,18 +3,130 @@ title: Versioning
...
+If not using the recommended syntax, the mapping proceeds by treating the first dot-separated fragment 
+as the qualifer and converts unsupported characters in a qualifier to an underscore;
+thus `1.x` becomes `1.0.0.x`, `v1` becomes `0.0.0.v1`, and `"1.0.0-v1.1"` becomes `"1.0.0.v1_1"` 
+hence the bundle replacement noted above.
+
+If you are creating an OSGi `MANIFEST.MF` for a bundle that also contains a `catalog.bom`, 
+you will need to use the mapped result (OSGi version syntax) in the manifest,
+but should continue to use the Brooklyn-recommended syntax in the `catalog.bom`.
+ 
+For those who are curious, the reason for the Brooklyn version syntax is to reconcile
```
Comment (geomacy) [inline guide/blueprints/catalog/versioning.md]: 👍  a great explanation to have

File: guide/blueprints/catalog/versioning.md [hunk 5]
```diff
@@ -3,18 +3,130 @@ title: Versioning
...
+When taking an ordering, for instance to find the highest version, 
+snapshot versions are always considered lower than release versions.
+Next, the natural order is taken on the major, minor, and patch fields.
+Next, a version with no qualifier is considered higher than one with a qualifier.
+Finally, the qualifier is taken in natural order.
+
+The natural order here is defined as ASCII-lexicographic comparison
+for any non-numeric segments (`"a" < "b"`) but numeric order for digit sequences 
+(`"9" < "10"`), so it does what is usually expected for versions (`1.9` < `1.10`
+and `"1.1-rc9-b" < "1.1-rc10-a"`).
```
Comment (ahgittin) [inline guide/blueprints/catalog/versioning.md]: Will make this difference explicit -- as noted in https://github.com/apache/brooklyn-server/pull/740 I think this is the right thing to do.

**Comment by ahgittin** [comment]
Thanks for great review comments.  All but one incorporated, including discussion about equality @geomacy .

The one that isn't is the statement about it "may" use co-bundling for searching.  This is because we want to get there but currently aren't, or mostly aren't, so being precise here is hard and might become stale.  Hopefully we'll soon _always_ do it and can update that word in the docs!

**Comment by ahgittin** [comment]
As https://github.com/apache/brooklyn-server/pull/743 is merged I'm merging this.

**Pull request merged**